### PR TITLE
Traceable nils

### PR DESCRIPTION
**Pull request by chrisseaton**
This PR sketches out an option for allowing users to trace where `nil` values have come from. For example:

```ruby
def read_from_hash(hash, key)
  hash[key]
end

def foo(hash)
  read_from_hash(hash, :baz)
end

hash = {foo: 14, bar: 2}

# We'll get a NME here - should say nil came from the hash's default value
# in #read_from_hash and was returned to here via foo
p foo(hash) * 2

# demo-hash.rb:13:in `<main>': undefined method `*' for nil:NilClass (NoMethodError)
# nil trace:
#         default hash value
#         returned from demo-hash.rb:2 in Object#read_from_hash
#         returned from demo-hash.rb:6 in Object#foo
```

This has an overhead when disabled of 2% in the interpreter and nothing in peak, and an overhead when tracing of 24% in the interpreter and 6% in peak. Enough to always run locally and when running test suites.

It works by allowing multiple instances of `nil`. When tracing, a log is created for control flow involving the `nil` instances. The log is a linked list of references to events and the nodes where they happened, so allocation has a simple always-fast-path, and as all tracing code is visible to PE, `nils` and their traces that don't escape are optimised away.

I'd need to turn the option into a proper option, add some more trace events, fix the failing PE tests (some complex things that were PE constant before aren't now, haven't investigated), and write some more elaborate demos to check it's actually useful for bug fixing.

Review by commit - the first two are large refactorings. 54cd3883038355dee6cc6a40b0fa7e06deb8f87a is the actual work.

Rubinius had `nil` tracing, but I think their ideas was only to record the original source of the nil, not where it went since, and I'm not sure how much was actually implemented.

* https://medium.com/metalanguage/nil-is-not-null-and-other-tales-febf6730f216
* https://twitter.com/rubinius/status/685534200213061632
* https://github.com/rubinius/rubinius/commit/da81b5daf4bf4633944caef8b898762eabada5ea

### CODE REVIEW

**Comment by eregon** [comment]
How did you measure the overhead? On a micro benchmark from that example? I'd think the overhead is higher if the nil is further away from the exception.

How much slower is jt test fast (in jvm configuration) with this option?

I wonder if just saving the SourceSection of the allocation wouldn't provide similar value, but without most of the costs.

And I wonder of the value in general, is tracing a nil manually difficult or time consuming? I've debugged many issues but I don't recall tracing nil's being a significant part of the debugging time.

**Comment by eregon** [comment]
> an overhead when tracing of 24% in the interpreter and 6% in peak. Enough to always run locally and when running test suites.

I disagree, 24% overhead when testing with the interpreter is way too much to be on by default.

**Comment by chrisseaton** [comment]
I mean people could turn it on by default for their specs, not on by default in TruffleRuby.

**Comment by chrisseaton** [comment]
All the commits up to 54cd3883038355dee6cc6a40b0fa7e06deb8f87a have zero overhead (tested using `test fast` and optcarrot.)

The commits to add tracing, but not enable it, have a 2% overhead on optcarrot, can't be measured on `test fast`.

Enabling tracing has a 24% overhead on `test fast`, 43% overhead on `optcarrot` interpreter, 6% overhead on `optcarrot` compiled.

**Comment by chrisseaton** [comment]
I log the node - seems cheaper than the source section, which may require walking up nodes to find one.

**Comment by nirvdrum** [comment]
I've run into issues in production Rails apps where a request results in a `NoMethodError` in a path where no one ever expected a `nil` value. If the value comes in as part of a request, it's not possible to reproduce unless you log the original request as well, so debugging those cases can be very annoying. Especially with usages of `&.` and Rails's `presence` call to skip over nil values.

**Comment by eregon** [comment]
I think this makes the code singificantly more verbose, and for instance I think `nil` -> `nil()` (especially when there was significant work to be able to just use a singleton, as needed for AST sharing) and other transformations like `r == Nil.INSTANCE` -> `Nil.is(result)` hurt readability and add cognitive overhead (i.e., what do these methods do, why are they even there, etc).
It's also going to cause endless conflicts with the TruffleString PR and other large PRs, which I see as a main blocker (at least at this time).
If it's a proper language option it won't be able to fold in interpreter (only in PE) and it will cost something in interpreter on native image. That is too valuable to sacrifice for a rarely-needed feature, even if it's just 2% (likely much more if using a proper option, would be intersting to check the actual overhead). If it's just a static final system property then it won't be possible to enable it on native image without rebuilding so it'd basically a JVM-only feature.

In summary, I think it's an interesting experiment, but I think it is not worth the cost in interpreter performance, readability, maintainability, and conflicts with other more fundamental necessary work.
